src/MonitorProfiler/Communication/IpcCommClient.cpp

```diff
@@ -0,0 +1,106 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "IpcCommClient.h"
+#include <memory>
+
+HRESULT IpcCommClient::Receive(IpcMessage& message)
+{
+    if (_shutdown.load())
+    {
+        return E_UNEXPECTED;
+    }
+    if (!_socket.Valid())
+    {
+        return E_UNEXPECTED;
+    }
+
+    //CONSIDER It is generally more performant to read and buffer larger chunks, in this case we are not expecting very frequent communication.
+    char buffer[sizeof(MessageType) + sizeof(int)];
+    int read = 0;
+    int offset = 0;
+
+    do
+    {
+        int read = recv(_socket, &buffer[offset], sizeof(buffer) - offset, 0);
+        if (read == 0)
+        {
+            return E_ABORT;
+        }
+        if (read < 0)
+        {
+#if TARGET_UNIX
+            if (errno == EINTR)
+            {
+                //Signal can interrupt operations. Try again.
+                continue;
+            }
+#endif
+            return SocketWrapper::GetSocketError();
+        }
+        offset += read;
+
+    } while (offset < sizeof(buffer));
+
+    message.MessageType = *reinterpret_cast<MessageType*>(buffer);
+    message.Parameters = *reinterpret_cast<int*>(&buffer[sizeof(MessageType)]);
+
+    return S_OK;
+}
+
+HRESULT IpcCommClient::Send(const IpcMessage& message)
+{
+    if (_shutdown.load())
+    {
+        return E_UNEXPECTED;
+    }
+    if (!_socket.Valid())
+    {
+        return E_UNEXPECTED;
+    }
+
+    char buffer[sizeof(MessageType) + sizeof(int)];
+    *reinterpret_cast<MessageType*>(buffer) = message.MessageType;
+    *reinterpret_cast<int*>(&buffer[sizeof(MessageType)]) = message.Parameters;
+
+    int sent = 0;
+    do
+    {
+        sent = send(_socket, buffer, sizeof(buffer), 0);
+
+        if (sent == 0)
+        {
+            return E_ABORT;
+        }
+        if (sent < 0)
+        {
+#if TARGET_UNIX
+            if (errno == EINTR)
+            {
+                //Signal can interrupt operations. Try again.
+                continue;
+            }
+#endif
+            return SocketWrapper::GetSocketError();
+        }
+    } while (sent < 0);
+
+    return S_OK;
+}
+
+void IpcCommClient::Shutdown()
+{
+    _shutdown.store(true);
+    shutdown(_socket,
```

Comment:
Should the return value be checked if its not 0 and then return `SocketWrapper::GetSocketError()`? Requires `Shutdown` to be updated to have an `HRESULT` return type.

Reply:
I can push the error handling to the caller, although I am unlikely to check the error code anyway since my action is to close the socket descriptor either way.

Reply:
Sure, but it's good to be able to log an error when a failure does occur. I want us to get logging on bad results as much as possible.